### PR TITLE
Allow to import the users from the more recently accessed Linux installation on the system

### PR DESCRIPTION
**Pull request by dgdavid**
> Related to bsc#1048983, allows to import, during the installation, users
from the **most recently accessed** Linux installation on the system
(according to `atime`).

---

See also https://trello.com/c/ECIP6QMV

### CODE REVIEW

File: test/lib/users/dialogs/inst_user_first.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        dialog.choose_users_handler
+      end
+    end
+
+    context "when there are multiple Linux installations" do
+      before do
+        # Mock access time: files in root2 are more recent than files in root3
+        allow(File).to receive(:atime) do |path|
+          path =~ /root2/ ? Time.now : Time.now - 1200
```
Comment (joseivanlopez) [inline test/lib/users/dialogs/inst_user_first.rb]: I am not sure if this is robust enough. I would use a specific time object:

```suggestion
          path =~ /root2/ ? Time.new(2018) : Time.new(2017)
```
Moreover, I would mock it in a different way:

```
allow(File).to receive(:atime).with(/root2/).and_return(Time.new(2018))
allow(File).to receive(:atime).with(/root3/).and_return(Time.new(2017))
```

File: test/lib/users/users_database_test.rb [hunk 2]
```diff
@@ -21,66 +21,102 @@
...
+  let(:fixture_root) { FIXTURES_PATH.join("root") }
+  let(:fixture_root2) { FIXTURES_PATH.join("root2") }
+  let(:fixture_root3) { FIXTURES_PATH.join("root3") }
+
+  before do
+    Users::UsersDatabase.all.clear
+
+    # Mock access time: files in root2 are more recent than files in root and root3
+    allow(File).to receive(:atime) do |path|
+      path =~ /root2/ ? Time.now : Time.now - 1200
```
Comment (joseivanlopez) [inline test/lib/users/users_database_test.rb]: Same comment as before. In fact, with this code is not clear which is most recent, root or root3?

File: test/lib/users/users_database_test.rb [hunk 3]
```diff
@@ -21,66 +21,102 @@
...
   end
 
   describe "#read_files" do
+    let(:passwd_atime) { Time.now }
+    let(:read_files) { subject.read_files(fixture_root2.join("etc")) }
+
     before do
       # Mock access time: passwd is more recent than shadow
       allow(File).to receive(:atime) do |path|
         path =~ /passwd$/ ? passwd_atime : passwd_atime - 1200
```
Comment (joseivanlopez) [inline test/lib/users/users_database_test.rb]: Please, use fixed time values.

File: test/lib/users/users_database_test.rb [hunk 4]
```diff
@@ -21,66 +21,102 @@
...
+
+      Users::UsersDatabase.import(fixture_root.join("etc"))
       Users::UsersDatabase.import("/nonexistent")
-      expect(Users::UsersDatabase.all).to be_empty
+      expect(databases).to be_empty
     end
   end
 
   describe "#read_files" do
+    let(:passwd_atime) { Time.now }
```
Comment (joseivanlopez) [inline test/lib/users/users_database_test.rb]: I would use a fixed time.

File: test/lib/users/dialogs/inst_user_first.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+describe Yast::InstUserFirstDialog do
+  subject(:dialog) { described_class.new }
+
+  describe "#choose_users_handler" do
+    let(:users_to_import_dialog) { instance_double(Yast::UsersToImportDialog, run: nil) }
+    let(:root2_path) { FIXTURES_PATH.join("root2") }
+    let(:root3_path) { FIXTURES_PATH.join("root3") }
+
+    context "when there is another Linux installation" do
```
Comment (joseivanlopez) [inline test/lib/users/dialogs/inst_user_first.rb]: NP: I would use "when there is a previous Linux installation"

File: test/lib/users/users_database_test.rb [hunk 6]
```diff
@@ -21,66 +21,102 @@
...
-      expect(databases.size).to eq 2
-      expect(databases.first.passwd).to start_with "a_user"
+      databases.clear
+      expect(databases).to be_empty
+
+      Users::UsersDatabase.import(fixture_root2)
+      Users::UsersDatabase.import(fixture_root3)
+      Users::UsersDatabase.import(fixture_root)
+      expect(databases.size).to eq(3)
+      expect(databases.first.passwd).to start_with("a_user")
```
Comment (joseivanlopez) [inline test/lib/users/users_database_test.rb]: In both cases the most recent is the same. I would try to do an example where the result is different.

File: src/lib/users/dialogs/inst_user_first.rb [hunk 7]
```diff
@@ -434,22 +434,26 @@ def process_import_form
...
+
           if UsersSimple.ReadUserData(dir)
-            # GetImportedUsers should not return nil, but better be safe
-            imported = UsersSimple.GetImportedUsers("local")
-            users = imported unless imported.nil?
-            break unless users.empty?
+            imported_users = UsersSimple.GetImportedUsers("local")
+
+            # UsersSimple.GetImportedUsers should not return nil, but better be safe
+            return imported_users unless imported_users.nil? || imported_users.empty?
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply src/lib/users/dialogs/inst_user_first.rb]: Not _Maybe_. It is definitely more readable. Great catch. Thank you!

File: test/lib/users/dialogs/inst_user_first.rb (hunk 1 above)
Comment (dgdavid) [reply test/lib/users/dialogs/inst_user_first.rb]: I must admit that this pieces of code were directly copy/paste from the existing one, without pay too much attention. But I agree with you, thank you Iván.

File: test/lib/users/users_database_test.rb [hunk 8]
```diff
@@ -21,66 +21,102 @@
...
+      Users::UsersDatabase.import(fixture_root)
+      Users::UsersDatabase.import(fixture_root2)
+      Users::UsersDatabase.import(fixture_root3)
+    end
+
+    it "returns found users databases" do
+      expect(databases.size).to eq(3)
+    end
+
+    it "returns the most recent accesed first" do
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply test/lib/users/users_database_test.rb]: There are already three databases. I'll add a test checking that they are correctly sorted. Thanks!

**Pull request merged**